[documentation] Generate images automatically in docs build, including them in our Sphinx docs

VindeeR opened the pull request:
Closes #2069
Closes #1160 

Improving the quality of the documentation generating it in an automated way ensuring images to always be up to date with instruction from provided

It uses our `runcmd` to execute `autosubmit` commands (create, run, etc.), placing images in a folder. Then, it moves them to the `_static`, and uses the `figure` directive to produce the `<figure>` used in the final docs. But, instead of injecting the `figure`, we get its HTML output, and send a `raw` directive.

This way, we control when the image will be used. Otherwise, `figure` would fail, saying that the image does not exist, as the images exist in the build, and not in the source directory, failing the `sphinx-build` execution.

An example of another directive that does something similar, used as reference: https://github.com/matplotlib/matplotlib/blob/799f26066b5b110fe7eee841ad979aee90238234/lib/matplotlib/sphinxext/plot_directive.py (although they do something more complicated, rendering a Jinja2 template, and then injecting values into the Sphinx state machine…

<!--

Thanks for your contribution! Please:
* List any related issues with a "closes" or "addresses" tag.
* Add a helpful title & description.
* Complete the checklist.

-->

**Check List**

- [x] I have read `CONTRIBUTING.md`.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [ ] Applied any dependency changes to `pyproject.toml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included in `CHANGELOG.md` if this is a change that can affect users.
- [x] Documentation updated.
- [x] If this is a bug fix, PR should include a link to the issue (e.g. `Closes #1234`).

kinow reviewed (commented):
We can simplify the part about needing to have an empty image file and calling two directives (`runcmd` followed by `figure`) by creating a new directive that does both, runs a command and produces a figure in the document.

The idea starts with something like this:

```diff
diff --git a/docs/source/userguide/defining_workflows/index.rst b/docs/source/userguide/defining_workflows/index.rst
index 46276c56..81df2d0f 100644
--- a/docs/source/userguide/defining_workflows/index.rst
+++ b/docs/source/userguide/defining_workflows/index.rst
@@ -37,18 +37,19 @@ be finished before launching the job that has the DEPENDENCIES attribute.
       FILE: two.sh
       DEPENDENCIES: One
 
+The resulting workflow can be seen in Figure :numref:`simple`
 
-.. runcmd:: find /home/docs/autosubmit/a000/plot/ -type f -iname "a000_*.png" -exec mv -- {} ./userguide/defining_workflows/fig/simple.png \;
-    :silent-output: 1
-    :prompt:
 
-The resulting workflow can be seen in Figure :numref:`simple`
+.. autosubmitfigure::
+    :command: create
+    :expid: a000
+    :type: png
+    :figure: fig/simple.png
+    :name: simple
+    :width: 100%
+    :align: center
+    :alt:  simple workflow plot
 
-.. figure:: fig/simple.png
-   :name: simple
-   :width: 100%
-   :align: center
-   :alt: simple workflow plot
 
    Example showing a simple workflow with two sequential jobs
```

The `runcmd` you already know. The `figure` is this (two links, one for figure, one for the docutils figure, which is called from the former).

* https://github.com/sphinx-doc/sphinx/blob/7b4164a5f2b64781475e64daf2d05cce2a0261d8/sphinx/directives/patches.py#L33
* https://github.com/docutils/docutils/blob/master/docutils/docutils/parsers/rst/directives/images.py

The goal here is do re-use the runcmd (either calling inside the new directive, or isolating the code in a function and importing in both directives), and reuse/call the figure.

I think the main challenge will be to configure the directive to write the image in the right location, which will be used when Sphinx produces the HTML file.

In case you need to debug it, here's an example setting for debugging Sphinx directives in PyCharm, calling `sphinx-build` (it will stop the breakpoint for every call of the directive, but you can cancel the execution after the first hit):

<img width="1518" height="1001" alt="image" src="https://github.com/user-attachments/assets/873693c2-4b14-4c48-bd42-508db8ac9f37" />

kinow reviewed (commented):
Example of how to return a different node (which comes from docutils, used to create the final document -- a document with nodes).

When silent is enabled, we can return an empty text block (first I thought of using, but we can use another one if you find an better option).

```diff
diff --git a/docs/source/ext/runcmd.py b/docs/source/ext/runcmd.py
index 058c63a2..fc8b8b4e 100644
--- a/docs/source/ext/runcmd.py
+++ b/docs/source/ext/runcmd.py
@@ -4,9 +4,9 @@ import re
 import shlex
 import subprocess
 import sys
-
 from pathlib import Path
 
+from docutils import nodes
 from docutils.parsers.rst import directives
 from sphinx.directives import code
 
@@ -169,7 +169,7 @@ class RunCmdDirective(code.CodeBlock):
 
         # silence the output if required
         if silent_output > 0:
-            output = ""
+            return [nodes.Text('')]
 
         # Add the prompt to our output if required
         if 'prompt' not in self.options:
```

kinow commented on docs/Makefile:
```diff
@@ -20,9 +20,51 @@ ALLSPHINXOPTS   = -d $(BUILDDIR)/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) sou
+# `.EXPORT_ALL_VARIABLES`, which was also made phony. All variables defined here
+# are exported to other targets. Use this feature with care!
+as_setup: .EXPORT_ALL_VARIABLES
+	$(eval AUTOSUBMIT_DIRECTORY := $(shell realpath -m $(BUILDDIR)/autosubmit))
```
Something like `$(shell realpath '.')/$(BUILDDIR)/autosubmit` might work? You can test locally...

kinow reviewed (commented):
Looking really good overall. Left a few comments after reading the code, once we are done with these review comments I'll have a look at the examples & generated site, and then it should be good to go :tada: Thanks Erick!!

kinow commented:
Git conflicts that need to be fixed too. Maybe you can squash the commits as well after the rebase.

kinow commented:
Oh, and I was going to switch to another branch to review another pull request, when I noticed the amount of changes in Git :sweat_smile: 

<img width="739" height="535" alt="image" src="https://github.com/user-attachments/assets/2b8f3029-09f9-429f-a129-88843cc75aa5" />

We should fix this too before merging, or at least try harder to avoid the files to be created in the source directory (clean solution is to output only in build directory, dirty one is to remove from source after build... kind of meh but could work).

kinow commented on test/integration/platforms/test_slurmplatform.py:
```diff
@@ -302,7 +302,7 @@ def test_run_simple_workflow_slurm(autosubmit_exp: AutosubmitExperimentFixture,
                     'SIMVH-1': {},
                 },
-                'SCRIPT': 'echo "0"',
+                'SCRIPT': 'sleep 0',
```
Looks great then!

VindeeR commented on test/integration/platforms/test_slurmplatform.py:
```diff
@@ -360,3 +360,262 @@ def test_run_all_wrappers_workflow_slurm(autosubmit_exp: AutosubmitExperimentFix
+    }
+
+    exp.autosubmit._check_ownership_and_set_last_command(exp.as_conf, exp.expid, 'run')
+    assert 0 == exp.autosubmit.run_experiment(_EXPID)
```
Those were examples not covered on the other tests.

The documentation have them but I had not create them in the other tests.

I can create a different branch for them, but as I went through the documentation this example of wrapper wasn't being covered.

kinow replied:
I think you explained it before, now I'm recLling. Thanks for adding the test then. Marking as resolved.

VindeeR commented on docs/Makefile:
```diff
@@ -20,9 +20,54 @@ ALLSPHINXOPTS   = -d $(BUILDDIR)/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) sou
+	AUTOSUBMIT_CONFIGURATION=$(AUTOSUBMIT_CONFIGURATION) autosubmit install
+	@echo "Autosubmit installed!"
+	autosubmit -lc DEBUG expid -dm -H "local" -d "Tutorial"
+	@echo "Autosubmit Experiment Created!"
```
I tried to create at the `readthedocs` after the `Makefile` but it didn't get the variables properly, I don´t remember trying inside the documentation, but I believe it could happen the same thing, that´s why I set there, where the environment variables where pointing to the right `path` before the possibility of being overwritten.

I too tried inside the `HTML` tag inside the `Makefile`, and it didn't work, that's why I believe this is the best place to initialize the experiment

VindeeR commented on docs/source/ext/autosubmitfigure.py:
```diff
@@ -0,0 +1,106 @@
+            path_from = f"{self.env.srcdir}/{self.options.get('path')}/code/exp_{self.options.get('name')}.yml"
+            path_to = f"{self.env.app.outdir.parent}/autosubmit/autosubmit/a000/conf/expdef_a000.yml"
+            if Path(path_from).is_file() and Path(path_to).is_file():
+                shutil.copy(path_from, path_to)
```
Actually, one is overwriting the the `jobs_<EXPID>.yml` and the other `expid_<EXPID>.yml` I should've make it a `for` to make it simpler to be honest

VindeeR commented on docs/source/userguide/defining_workflows/index.rst:
```diff
@@ -32,15 +31,20 @@ be finished before launching the job that has the DEPENDENCIES attribute.
-   :alt: simple workflow plot
 
-   Example showing a simple workflow with two sequential jobs
+The resulting workflow can be seen in Figure
```
I've done a few tests to double-check but this refs are lost by using the custom Directive

kinow commented:
Hmmm, works locally. On RTD it still needs to handle the basepath `/en/2465/`, for example, in this PR.

kinow commented:
Slurm failed, but the fix in already in another PR. Kicking the build...

kinow commented:
(oh, rebased too)

kinow commented on docs/source/troubleshooting/changelog.rst:
```diff
@@ -523,18 +523,24 @@ New format
+    :name: new_dependencies
+    :width: 100%
+    :align: center
+    :alt: new_dependencies
```
Note: path is not needd, and the `figure` has only the file name... we are moving to somewhere like `build/html/_static/troubleshooting/new_dependencies.png`... so images must be unique per file, but if that becomes a problem, later we can include the `rst` file name in the target path.

VindeeR commented:
> There you go, @VindeeR .
> 
> Have a look at my commit. I didn't squash it so you could review it, but feel free to squash it later. I've rebased it.
> 
> I moved the images to the build directory (it's a recommendation of Sphinx to never mess with source code -- I had a bug with other directives when messing with source files, as even if it's `gitignore`'d, the files are still accessed/available to directives).
> 
> I had to make a few changes in the directive so that it rendered correctly, like using `raw` + `figure` instead of `figure` (`figure` renders, but instead of complaining about the image, I ask it to please shut up and give me the content, which I insert into a raw as HTML 😆 ).
> 
> I also moved some code to functions, so it's easier to cover with more tests later, and left a TODO about testing this with other builders like PDF (not important nor urgent right now, it's more important to have an up to date documentation).
> 
> Locally, you can build it with `make html`, but the images won't work in the browser using `file://`. You can test with `cd docs/` and then `python -d build/html 8000` and check the result in `http://localhost:8000`.
> 
> I haven't tested `make watch but if RTD and local works, we should focus on that first, and worry about the rest later (having good & up to date docs is more important).
> 
> I'm leaving this approved as I'll go on holidays, @VindeeR , but please do a second review so you confirm my changes don't have anything weird, as it's late here 🙂
> 
> Cheers

Okay I just finished local testing/reviewing/rebasing everything worked perfectly fine, I just had to remove a test that was left behind other than that everything was perfect to go. 

Thanks for the throughout review, and some improvements on the code

kinow commented:
This is a huge improvement to our docs, @VindeeR ! Thanks a lot for working on this!!! :tada: :star: :partying_face:
